core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -239,9 +239,9 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
     nodeChangeHandlers.foreach(zkClient.registerZNodeChangeHandlerAndCheckExistence)
 
     info("Deleting log dir event notifications")
-    zkClient.deleteLogDirEventNotifications()
+    zkClient.deleteLogDirEventNotifications(controllerContext.epochZkVersion)
```

Comment:
Thanks for pointing this out. This is indeed a very dangerous race condition. If it happens, the current controller (Broker B) cannot update any zookeeper state due to controller epoch zkVersion mismatch and no other broker can become the controller because the current controller (Broker B) does not release the "lock" for `\controller` znode.

Wrapping `\controller` creation and `\controller_epoch` update in a zookeeper transaction can prevent this race condition and I think it is a safe option. I will make the change and see whether there will be performance overhead in the perf testing.

Reply:
Hey @hzxa21, it seems that what you and Jun suggested to do is to have a single multiops that 1) updates controller path, 2) read controller epoch and 3) updates controller epoch. Another alternative approach is to have a single multiops that 1) updates controller path and 2) reads controller epoch with its zkversoin. Then the controller can updates controller epoch with the addition zkversion check.

Do you think the alternative approach would avoid the race condition and ensure correctness? If so, I am wondering if the alternative would be easier to reason about. I find it a bit easier because the it follows the idea that all zookeeper write operation by controller will be based on the controller epoch zkversion check, except for the controller znode write operation which by design can not rely on the controller epoch zkversion check. And a multiop that does one write and one read seems simpler than a multiop that does write-read-write.

Reply:
@lindong28 From a design and code readability perspective, I agree with what you have proposed (First atomic read `\controller_epoch` and create `\controller`, then update `\controller_epoch`). From the implementation perspective, zookeeper does not have a `read` Op meaning that we cannot perform `read` operation with the `multi` (see http://people.apache.org/~larsgeorge/zookeeper-1215258/build/docs/dev-api/org/apache/zookeeper/Op.html).

Basically, we use the time when a broker succeeds in incrementing the controller epoch as the "commit" point of the controller election and use the time when a broker succeeds in creating `\controller` znode as the "prepare" point. So for the correctness of the controller election "commit", we need to ensure `\controller_epoch` doesn't change from "prepare" to "commit". To achieve, we can implement the logic using zk `multi` following the steps:
1. Read `\controller_epoch` to get the current controller epoch **e1** with zkVersion **v1**
2. Create `\controller` if `\controller_epoch` zkVersion matches **v1** (use zk `multi`)
3. Update `\controller_epoch` to be **e1+1** if its zkVersion matches **v1** (zk conditional set)

Reply:
PR updated to address this issue.